cilium/CFP-28175-BGP-v2-API-changes.md

```diff
@@ -0,0 +1,895 @@
+* Router-ID must be unique per BGP instance. We cannot set it via CiliumBGPClusterConfig as it is a cluster-wide configuration.
+    * Today we detect local IP and configure it. But this is error prone as multiple BGP instances will pick the same router-id.
+    * Operators may want to set router-id explicitly.
+* Local-Address used in peering configuration is also node-specific configuration. In multihoming scenarios, we may want to force peering over specific interfaces.
+
+The following configuration options should be supported per ASN:
+* Router ID
+* Global local port to be configured
+* Local Address to be used for given neighbor
+* Local Port to be used for given neighbor
+
+
+## Configuration override via CRD CiliumBGPNodeConfigOverride
+
+We can specify node-specific configuration via CiliumBGPNodeConfigOverride. The operator will consume this resource and update the CiliumBGPNodeConfig resource.
+
+```go
+// CiliumBGPNodeConfigOverride is used to overrides some of the BGP configurations which are node local.
+type CiliumBGPNodeConfigOverride struct {
+	// +deepequal-gen=false
+	metav1.TypeMeta `json:",inline"`
+	// +deepequal-gen=false
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec CiliumBGPNodeConfigOverrideSpec `json:"spec"`
+}
+
+type CiliumBGPNodeConfigOverrideSpec struct {
+	// NodeRef is the name of the node for which the BGP configuration is overridden.
+	//
+	// +kubebuilder:validation:Required
+	NodeRef string `json:"nodeRef"`
+
+	// +listType=map
+	// +listMapKey=localASN
+	BGPInstances []CiliumBGPNodeConfigInstanceOverride `json:"bgpInstances"`
+}
+
+type CiliumBGPNodeConfigInstanceOverride struct {
+	// LocalASN is used to identify the BGP instance to override.
+	// +kubebuilder:validation:Required
+	LocalASN uint32 `json:"localASN"`
+
+	// RouterID is BGP router id to use for this instance. It must be unique across all BGP instances.
+	// If not specified, the node's IP address is used.
+	//
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Format=ipv4
+	RouterID string `json:"routerID,omitempty"`
+
+	// LocalPort is port to use for this instance.
+	// +kubebuilder:validation:Optional
+	LocalPort int32 `json:"localPort,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	Peers []CiliumBGPNodeConfigPeerOverride `json:"peers,omitempty"`
+}
+
+type CiliumBGPNodeConfigPeerOverride struct {
+	// PeerAddress is used to identify the peer to override.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Format=ip
+	PeerAddress string `json:"peerAddress"`
+
+	// LocalAddress is the IP address to use for this peer.
+	// +kubebuilder:validation:Optional
+	LocalAddress string `json:"localAddress,omitempty"`
+
+	// LocalPort is port to use for this peer.
+	// +kubebuilder:validation:Optional
+	LocalPort int32 `json:"localPort,omitempty"`
+}
+```
+
+## Backward Compatibility
```

Comment:
This section should be removed, we are reconsidering how migration from v1->v2 will happen. Alternate approach is something like this 
- Provide a CLI which consumes CiliumBGPPeeringPolicy and outputs equivalent BGPv2 CRDs.
- New CRDs can be applied by user
- Enable BGPv2 via feature flag - helm upgrade.